operator/v1/types_dns.go

```diff
@@ -203,6 +254,43 @@ type UpstreamResolvers struct {
 	// +optional
 	// +kubebuilder:default="Sequential"
 	Policy ForwardingPolicy `json:"policy,omitempty"`
+
+	// transport allows cluster administrators to opt-in to using a DNS-over-TLS
+	// connection between cluster DNS and an upstream resolver(s). Configuring
+	// TLS as the transport at this level without configuring a CABundle will
+	// result in the system certificates being used to verify the serving
+	// certificate of the upstream resolver(s).
+	//
+	// Possible values:
+	// "" (empty) - This is the default and results in DNS queries being sent over cleartext.
+	// "cleartext" - Cluster admin specified cleartext option. This results in the same functionality
+	// as an empty value but may be useful when a cluster admin wants to be more explicit about the transport,
+	// or wants to switch from "tls" to "cleartext" explicitly.
+	// "tls" - This indicates that DNS queries should be sent over a TLS connection. If Transport is set to TLS,
+	// you MUST also set ServerName.
+	//
+	// +optional
+	Transport DNSTransport `json:"transport,omitempty"`
+
+	// caBundle references a ConfigMap that must contain either a single
+	// CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+	// by different CAs). This allows cluster administrators to provide their
+	// own CA or CA bundle for validating the certificate of upstream resolvers.
+	//
+	// 1. The configmap must contain a `cabundle.crt` key
+	// 2. The value must be a PEM encoded certificate or bundle.
+	// 3. The administrator must create this configmap in the openshift-config namespace.
+	//
+	// +optional
+	// +kubebuilder:validation:Optional
+	CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
```

Comment:
@deads2k I had a question for you on this one; I noticed that by default this will show up in `dns.operator.openshift.io/default` like this:
```
 spec:
  logLevel: Normal
  nodePlacement: {}
  operatorLogLevel: Normal
  upstreamResolvers:
    caBundle:
      name: ""
```
I have a slight concern that this will be confusing for users given that `caBundle` only applies when `Transport: tls`. Is it possible to "hide" this by default with `+nullable` and if so, is that advisable from an API perspective?

Reply:
Seeing available configuration is useful for people learning about what they can and cannot do.  In addition, we always want empty and missing to mean the same thing and avoiding pointers ensures that.   I would prefer to  keep the empty configuration over omiting it.

Reply:
> Seeing available configuration is useful for people learning about what they can and cannot do.

wouldn't that be an argument to never use omitempty anywhere?  also isn't "oc explain" the appropriate way to learn what you can/cannot do?  (especially since that also gives you descriptions of the fields)

>  In addition, we always want empty and missing to mean the same thing and avoiding pointers ensures that.

we sometimes want "empty" to mean "i set this to empty" and missing to mean "i didn't set this to anything, so take whatever default action you want on it", no?

not trying to debate anything about this specific choice here, but i'd like to better understand the guidelines you're describing because they don't align w/ my understanding

Reply:
@deads2k follow up question, `caBundle` on it's own won't do anything even if it's set to a value so should `Transport` and `ServerName` be made visible somehow to make the dependency more obvious?

Reply:
@bparees I think that might make sense; right now we're only allowing for configuration of a CA but there are other TLS params that customers may want to configure eventually (they're out of scope of this enhancement currently). @deads2k since you reviewed the enhancement proposal would you mind giving me a :+1: / :-1: on the suggested change?

Reply:
> @deads2k since you reviewed the enhancement proposal would you mind giving me a  /  on the suggested change?

Grouping sounds nice.